pkg/kafka/policy.go

```diff
@@ -227,6 +268,13 @@ func (req *RequestMessage) ruleMatches(rule api.PortRuleKafka) bool {
 		return matchOffsetFetchReq(val, rule)
 	case *proto.ConsumerMetadataReq:
 		return true
+	case nil:
+		// This is the case when requests like
```

Comment:
Please clarify that this covers the case where the Kafka parser is not aware of the request type

Reply:
I am not sure that's the right statement. Kafka parser is currently written to only further parse "topic" specific requests. This code is hit even if it understands the "apiKey" but since we have not parsed it further in the parser, the "req.Request" field is nil. This is true for requests like "heartbeat", "findCoordinator" et al. It understands the apiKey Kind, but we have not parsed it beyond that.

Reply:
Earlier this was solely working because we never tested with a rule have apiKey something other than fetch/produce and topic in the same rule, so this check in `ruleMatches` was making it work.

```
// If the rule contains no additional conditionals, it is not required
	// to match into the request specific fields.
	if rule.Topic == "" && rule.ClientID == "" {
		return true
	}
```

Reply:
Yes, that's why the check is in there.

The problem that we are discussing here is what should be the behaviour if a topic is specified but the kafka request in question, e.g. apiVersions, does not have a topic in the header. Right now, the user has to solve this by specifying two rules:
- rule 1 to whitelist all apiKeys that have topic fields and AND a match on the topic name
- rule 2 to whitelist all apiKeys required which do not have a topic field.

I assume the problem you ran into is that in order to properly implement roles, you need to allow apiKeys which do not have topic fields. Is that correct?

Reply:
Yes, That's correct hence the check in `matchNonTopicRequests` 
`if rule.Topic != "" && len(rule.APIKey) != 0 {`

Reply:
We are talking about the same issue in two threads, let's continue in the other thread